### PR TITLE
[libsodium] Migrate build system to CMake

### PR DESCRIPTION
**Pull request by myd7349**
See:

https://github.com/Microsoft/vcpkg/pull/4214#issuecomment-470484603
https://github.com/Microsoft/vcpkg/issues/3382
https://github.com/Microsoft/vcpkg/issues/3627

### CODE REVIEW

**Comment by past-due** [comment]
As long as I manually copy the CMakeLists.txt into the /src folder, this currently seem to _compile_ on macOS.

Looks like there might be a few flags / defines missing though, as the `install-*-out.log` is full of:
```
.../buildtrees/libsodium/src/1.0.17-2ff0ca6738/src/libsodium/include/sodium/private/common.h:14:3: warning: *** The library is being compiled using an undocumented method. [-W#warnings]
# warning *** The library is being compiled using an undocumented method.
  ^
.../buildtrees/libsodium/src/1.0.17-2ff0ca6738/src/libsodium/include/sodium/private/common.h:15:3: warning: This is not supported. It has not been tested, it might not [-W#warnings]
# warning This is not supported. It has not been tested, it might not
  ^
.../buildtrees/libsodium/src/1.0.17-2ff0ca6738/src/libsodium/include/sodium/private/common.h:16:3: warning: work as expected, and performance is likely to be suboptimal. [-W#warnings]
# warning work as expected, and performance is likely to be suboptimal.
  ^
3 warnings generated.
```
Hitting this in `src/libsodium/include/sodium/private/common.h`:
```c
#if !defined(_MSC_VER) && (!defined(CONFIGURED) || CONFIGURED != 1)
# warning *** The library is being compiled using an undocumented method.
# warning This is not supported. It has not been tested, it might not
# warning work as expected, and performance is likely to be suboptimal.
#endif
```

And the `install-*-err.log` contains:
```
/usr/bin/ranlib: file: .../vcpkg/packages/libsodium_x64-osx/lib/liblibsodium.a(blake2b-compress-avx2.c.o) has no symbols
/usr/bin/ranlib: file: .../vcpkg/packages/libsodium_x64-osx/lib/liblibsodium.a(blake2b-compress-sse41.c.o) has no symbols
/usr/bin/ranlib: file: .../vcpkg/packages/libsodium_x64-osx/lib/liblibsodium.a(blake2b-compress-ssse3.c.o) has no symbols
/usr/bin/ranlib: file: .../vcpkg/packages/libsodium_x64-osx/lib/liblibsodium.a(poly1305_sse2.c.o) has no symbols
/usr/bin/ranlib: file: .../vcpkg/packages/libsodium_x64-osx/lib/liblibsodium.a(argon2-fill-block-avx2.c.o) has no symbols
/usr/bin/ranlib: file: .../vcpkg/packages/libsodium_x64-osx/lib/liblibsodium.a(argon2-fill-block-avx512f.c.o) has no symbols
/usr/bin/ranlib: file: .../vcpkg/packages/libsodium_x64-osx/lib/liblibsodium.a(argon2-fill-block-ssse3.c.o) has no symbols
/usr/bin/ranlib: file: .../vcpkg/packages/libsodium_x64-osx/lib/liblibsodium.a(pwhash_scryptsalsa208sha256_sse.c.o) has no symbols
/usr/bin/ranlib: file: .../vcpkg/packages/libsodium_x64-osx/lib/liblibsodium.a(curve25519_sandy2x.c.o) has no symbols
/usr/bin/ranlib: file: .../vcpkg/packages/libsodium_x64-osx/lib/liblibsodium.a(fe51_invert.c.o) has no symbols
/usr/bin/ranlib: file: .../vcpkg/packages/libsodium_x64-osx/lib/liblibsodium.a(fe_frombytes_sandy2x.c.o) has no symbols
/usr/bin/ranlib: file: .../vcpkg/packages/libsodium_x64-osx/lib/liblibsodium.a(chacha20_dolbeau-avx2.c.o) has no symbols
/usr/bin/ranlib: file: .../vcpkg/packages/libsodium_x64-osx/lib/liblibsodium.a(chacha20_dolbeau-ssse3.c.o) has no symbols
/usr/bin/ranlib: file: .../vcpkg/packages/libsodium_x64-osx/lib/liblibsodium.a(salsa20_xmm6.c.o) has no symbols
/usr/bin/ranlib: file: .../vcpkg/packages/libsodium_x64-osx/lib/liblibsodium.a(salsa20_xmm6int-avx2.c.o) has no symbols
/usr/bin/ranlib: file: .../vcpkg/packages/libsodium_x64-osx/lib/liblibsodium.a(salsa20_xmm6int-sse2.c.o) has no symbols
/usr/bin/ranlib: file: .../vcpkg/packages/libsodium_x64-osx/lib/liblibsodium.a(randombytes_nativeclient.c.o) has no symbols
```

Looking at the `configure.ac`, it looks like the CMakeLists.txt could exclude at least:
`randombytes/nativeclient/randombytes_nativeclient.c`
if `__native_client__` is not defined.

File: ports/libsodium/portfile.cmake [hunk 1]
```diff
@@ -6,73 +6,36 @@ vcpkg_from_github(
...
 )
 
-if (VCPKG_LIBRARY_LINKAGE STREQUAL "dynamic")
-    set(LIBSODIUM_RELEASE_CONFIGURATION ReleaseDLL)
-    set(LIBSODIUM_DEBUG_CONFIGURATION DebugDLL)
-else()
-    set(LIBSODIUM_RELEASE_CONFIGURATION Release)
-    set(LIBSODIUM_DEBUG_CONFIGURATION Debug)
-endif()
+vcpkg_configure_cmake(
```
Comment (past-due) [inline ports/libsodium/portfile.cmake]: To copy the CMakeLists.txt to the SOURCE_PATH, something like the following:
```cmake
file(COPY
    ${CMAKE_CURRENT_LIST_DIR}/CMakeLists.txt
    DESTINATION ${SOURCE_PATH}
)
```
before the `vcpkg_configure_cmake` line works.

**Comment by Rastaban** [comment]
I am seeing the following new build breaks with the current commit on this PR:

```
 Processing x64-windows                    881 vs 884
     libsodium                     **regression**                Fail vs Pass
 Processing x64-windows-static             775 vs 778
     libsodium                     **regression**                Fail vs Pass
 Processing x86-windows                    863 vs 864
     libsodium                     **regression**                Fail vs Pass
```

**Comment by vicroms** [comment]
Building successfully on all triplets in our CI, even UWP 😮 

It would do good to investigate the issues mentioned by @past-due 

Unrelated, but looks like package `fizz` that depends on `libsodium` is still failing in OS X and Linux, but making `libsodium` portable to those platfomrs should be a first step to fix the `fizz` port.

File: ports/libsodium/portfile.cmake (hunk 1 above)
Comment (myd7349) [reply ports/libsodium/portfile.cmake]: @past-due Thanks for your kindly review.

I got this line lost when I tried to resolve the conflict after the first commit of this PR. It's my fault. 😄

**Comment by myd7349** [comment]
@past-due @Rastaban @vicroms Thanks for the test report and advice.

I will try to refine this PR in the following days.

**Comment by Rastaban** [comment]
@myd7349 what is the status of this PR?  It looks like it is passing all of the tests and has fixed the libsodium build failures.  Do you feel like it needs additional refining or is it ready to remove the WIP tag?

**Comment by myd7349** [comment]
@Rastaban My part is almost done. But I want to make sure that all tests passing. I will test it this weekend.

**Comment by myd7349** [comment]
Known issues:

* After running `dumpbin /EXPORTS libsodium.dll`, I got different results without and with this patch:
    * Using original MSBuild: [old.txt](https://github.com/Microsoft/vcpkg/files/3028916/old.txt)
    * Using this CMake build: [new.txt](https://github.com/Microsoft/vcpkg/files/3028919/new.txt)

**Comment by myd7349** [comment]
`libsodium` itself provides a cmake module named `Findsodium.cmake`: https://github.com/jedisct1/libsodium/blob/master/contrib/Findsodium.cmake

So I think it is reasonable to rename the project to `sodium`.

**Comment by myd7349** [comment]
Known issues:

* After running `dumpbin /EXPORTS libsodium.dll`, I found the exported symbols are slightly different from those just before this patch:
    * Using original MSBuild: [old.txt](https://github.com/Microsoft/vcpkg/files/3028916/old.txt)
    * Using this CMake build: [new.txt](https://github.com/Microsoft/vcpkg/files/3028919/new.txt)

**Comment by myd7349** [comment]
To run the tests:

* Copy `CMakeLists.txt` and `sodiumConfig.cmake.in` to `libsodium`'s source dir
* mkdir build
* cd build
* cmake .. -DBUILD_SHARED_LIBS=ON
* cmake --build . --config Release
* ctest -C Release

**Comment by myd7349** [comment]
Known issue:

* After running `dumpbin /EXPORTS libsodium.dll`, I found the exported symbols are slightly different from those just before this patch:
    * Using original MSBuild: [old.txt](https://github.com/Microsoft/vcpkg/files/3028916/old.txt)
    * Using this CMake build: [new.txt](https://github.com/Microsoft/vcpkg/files/3028919/new.txt)

**Comment by myd7349** [comment]
I think it is time to move forward. So I removed the `WIP` tag.
It is not perfect. But it is a good start.

**Pull request merged**